ring/partition_ring_model.go

```diff
@@ -0,0 +1,336 @@
+		out[partition.State]++
+	}
+
+	return out
+}
+
+// AddPartition adds a new partition to the ring. Tokens are auto-generated using the spread minimizing strategy
+// which generates deterministic unique tokens.
+func (m *PartitionRingDesc) AddPartition(id int32, state PartitionState, now time.Time) {
+	// Spread-minimizing token generator is deterministic unique-token generator for given id and zone.
+	// Partitions don't use zones.
+	spreadMinimizing := NewSpreadMinimizingTokenGeneratorForInstanceAndZoneID("", int(id), 0, false)
+
+	m.Partitions[id] = PartitionDesc{
+		Tokens:         spreadMinimizing.GenerateTokens(optimalTokensPerInstance, nil),
+		State:          state,
+		StateTimestamp: now.Unix(),
+	}
+}
+
+// UpdatePartitionState changes the state of a partition. Returns true if the state was changed,
+// or false if the update was a no-op.
+func (m *PartitionRingDesc) UpdatePartitionState(id int32, state PartitionState, now time.Time) bool {
+	d, ok := m.Partitions[id]
+	if !ok {
+		return false
+	}
+
+	if d.State == state {
+		return false
+	}
+
+	d.State = state
+	d.StateTimestamp = now.Unix()
+	m.Partitions[id] = d
+	return true
+}
+
+// RemovePartition removes a partition.
+func (m *PartitionRingDesc) RemovePartition(id int32) {
+	delete(m.Partitions, id)
+}
+
+// HasPartition returns whether a partition exists.
+func (m *PartitionRingDesc) HasPartition(id int32) bool {
+	_, ok := m.Partitions[id]
+	return ok
+}
+
+// AddOrUpdateOwner adds or updates a partition owner in the ring. Returns true, if the
+// owner was added or updated, false if it was left unchanged.
+func (m *PartitionRingDesc) AddOrUpdateOwner(id string, state OwnerState, ownedPartition int32, now time.Time) bool {
+	prev, ok := m.Owners[id]
+	updated := OwnerDesc{
+		State:          state,
+		OwnedPartition: ownedPartition,
+
+		// Preserve the previous timestamp so that we'll NOT compare it.
+		// Then, if we detect that the OwnerDesc should be updated, we'll
+		// also update the UpdateTimestamp.
+		UpdatedTimestamp: prev.UpdatedTimestamp,
+	}
+
+	if ok && prev.Equal(updated) {
+		return false
+	}
+
+	updated.UpdatedTimestamp = now.Unix()
+	m.Owners[id] = updated
+	return true
+}
+
+// RemoveOwner removes a partition owner.
+func (m *PartitionRingDesc) RemoveOwner(id string) {
+	delete(m.Owners, id)
+}
+
+// HasOwner returns whether a owner exists.
+func (m *PartitionRingDesc) HasOwner(id string) bool {
+	_, ok := m.Owners[id]
+	return ok
+}
+
+// Merge implements memberlist.Mergeable.
+func (m *PartitionRingDesc) Merge(mergeable memberlist.Mergeable, localCAS bool) (memberlist.Mergeable, error) {
+	return m.mergeWithTime(mergeable, localCAS, time.Now())
+}
+
+func (m *PartitionRingDesc) mergeWithTime(mergeable memberlist.Mergeable, localCAS bool, now time.Time) (memberlist.Mergeable, error) {
+	if mergeable == nil {
+		return nil, nil
+	}
+
+	other, ok := mergeable.(*PartitionRingDesc)
+	if !ok {
+		// This method only deals with non-nil rings.
```

Comment:
Removed both here and `Ring` because we actually handle `nil` in both cases: https://github.com/grafana/dskit/pull/474/commits/ca4eaff6260631a99a0ca06a09200c27faa94072